### DISCUSSION TITLE
Getting Module Not Found Error for next/server

### DISCUSSION THREAD

**Discussion by bwheeele** [nextauthjs/next-auth#10058] [Help]
I have been trying to use the new auth.js so I have   "next": "^14.1.0",
    "next-auth": "^5.0.0-beta.11", I am still using pages though as well. I am trying to migrate from using Next.js 13.4 and MSAL to using next-auth. I have an auth.ts, auth.config.ts, my nextauth api and a middleware. I keep getting this error though while trying to login or calling getSession() as it seems when my app hits the api/auth/error I get a 500 and the error reads:

Error [ERR_MODULE_NOT_FOUND]: Cannot find module '/node_modules/next/server' imported from /node_modules/next-auth/lib/env.js
Did you mean to import "next/server.js"?

I have tried looking everywhere but can't find any information about this error and what might be causing it. Hopefully I can find someone here who knows what the deal might be. Thank you!

**Comment by CommanderRedYT** [comment]
Did you get anything here yet?  I have this exact same error...

**Comment by santiagotri** [comment]
Same here, with same environment (pages architecture) "next-auth": "^5.0.0-beta.11"

**Comment by choosenname** [comment]
I fixed it by using "next": "14.1.4"

**Reply by CommanderRedYT** [reply]
> I fixed it by using "next": "14.1.4"

Had 14.2.3 before, you think it was broken after 14.1.4?

**Reply by choosenname** [reply]
Yap, 14.1.4 is the latest version that worked for me

**Reply by sKopheK** [reply]
> I fixed it by using "next": "14.1.4"

`14.2.16` here, i don't want to go back with next js because of an error in a library

**Comment by sKopheK** [comment]
i like the comment in `next-auth/lib/env.js` just before the failing import 
```
// @ts-expect-error Next.js does not yet correctly use the `package.json#exports` field
import { NextRequest } from "next/server";
```

**Reply by GNRSN** [reply]
I'm also having this issue, it works with next 15 but it fails in integration-tests run through vitest (vite).
> i like the comment in `next-auth/lib/env.js` just before the failing import
> 
> ```
> // @ts-expect-error Next.js does not yet correctly use the `package.json#exports` field
> import { NextRequest } from "next/server";
> ```

Is there an upstream github issue?

**Reply by rinarakaki** [reply]
Does this https://github.com/nextauthjs/next-auth/discussions/9385#discussioncomment-11064988 solve your problem?

**Reply by GNRSN** [reply]
It did, thanks!

**Comment by shahargl** [comment]
is there any workaround for that? why should I go back to 14.1.4 for that? :(

**Comment by yaman3bd** [comment]
I have the same error. Im currently migrating from Pages Router to App Router. My approach is to first upgrade Next.js and all related libraries, adapt to the new APIs, and ensure everything works as expected with the Pages Router. Once that’s stable, I’ll proceed with the migration to the App Router.

```bash
⨯ [Error: Cannot find module '/node_modules/next/server' imported from /node_modules/next-auth/lib/env.js
Did you mean to import next/server.js?] {
  code: 'ERR_MODULE_NOT_FOUND',
  url: 'file:///node_modules/next/server'
}
```
my dependencies:

```json
{
  "next": "15.0.3",
  "next-auth": "^5.0.0-beta.25"
}
```

**Comment by andrewdownss** [comment]
Has anyone found a work around?

**Reply by trainerkelly** [reply]
Hey there! I wrote about [the fix (workaround?)](https://github.com/nextauthjs/next-auth/discussions/10058#discussioncomment-11740620) I found below. Take a look and let me know if it solves your issues – it solved mine!

**Comment by trainerkelly** [comment]
Hey there! Thank you so much for posting this – I ran into the same error last night and for some reason, the formatting of the message in your post really clarified what the error was saying. I became determined to fix it and...I think I did! Or at least I have some sort of work-around. 

Please read this full post because I ran into some additional errors once the original one was fixed. 

### MODAL NOT FOUND Fix 
The error message specifies a file location and a suggested import. Go to the folder indicated after the "imported from" wording (in the case of the original post, it was /node_modules/next-auth/lib/). Open the following files: 

- env.js
- index.js
- actions.js 

_(Yes, the .js files – **NOT** the other ones ending in .ts(.map)!)_ 

Update the following imports by adding ".js" at the end of them: 

- Inside "env.js":
  - import { NextRequest } from "next/server";  --> import { NextRequest } from "next/server.js”; 
- Inside "index.js":
  - import { headers } from "next/headers"; --> import { headers } from "next/headers.js”;
  - import { NextResponse } from "next/server"; --> import { NextResponse } from "next/server.js”;
- Inside "actions.js": 
  - import { headers as nextHeaders, cookies } from "next/headers"; --> import { headers as nextHeaders, cookies } from "next/headers.js”; -
  - import { redirect } from "next/navigation"; --> import { redirect } from "next/navigation.”js; 

You can also search for the comment "ts-expect-error Next.js does not yet correctly use the `package.json#exports` field" because each path you need to update has that as the comment above it.

This should resolve that specific issue. 

That being said, I ran into two additional issues after I resolved this one that you may want to resolve first. 

### Additional Issues
####  ⨯ Error: Could not parse module '[project]/node_modules/next/dist/server/route-modules/app-route/vendored/contexts/app-router-context.js'
I'm happy to say that this is an easy enough fix – remove Turbopack (--turbo) from the next dev command in package.json. (Credit to a comment [on Reddit](https://www.reddit.com/r/nextjs/comments/1dcc8ti/next_auth_error_could_not_parse_module/?rdt=44988) for this. It does look like a [know issue](https://github.com/nextauthjs/next-auth/issues/11127#issuecomment-2276487943) that will be fixed sometime in the next release, so Turbopack doesn't have to stay off forever.) 

#### ⨯ TypeError: (0 , react__WEBPACK_IMPORTED_MODULE_1__.cache) is not a function at uncachedAuth (src/server/auth/index.ts:8:19)
`const auth = cache(uncachedAuth);` was throwing me that error. I resolved this error by uninstalling the React and React-Dom version I had installed (18.3.1) and reinstalling them to the latest version (19.0.0). It worked beautiful and fixed problem. (Credit to [a comment left on Stack Overflow](https://stackoverflow.com/questions/75849616/why-am-i-getting-a-cache-not-found-in-react-error-when-trying-to-import-cach) for this idea.)

Everything else ran as expected. 

I hope this is helpful to somebody! 

(If anyone else has a better solution, please post it! This took me about 6 hours total of searching Google, docs, using ChatGPT, and following the error messages in the command line to come up with this. I'm still new to this side of web dev, so there is a good chance I missed a much better solution. Thank you!)

**Reply by schumacher-jan** [reply]
Solved my solution. Thx!

**Reply by gompu123** [reply]
Thanks so much for sharing this solution! It worked perfectly for me, and I was able to fix the issue.

I applied it using `pnpm patch`, and it worked like a charm. Really appreciate your help and the time you put into sharing this.

```
pnpm patch next-auth@5.0.0-beta.25

pnpm patch-commit '/Users/giyeonkim/Projects/todaktalk/app/node_modules/.pnpm_patches/next-auth@5.0.0-beta.25'
```

```
diff --git a/src/lib/actions.ts b/src/lib/actions.ts
index 80081fa2e0603c0317ee7d9ea599d31e731e6b4d..d2063dc62070ba7119e7296b89503c4493b2c720 100644
--- a/src/lib/actions.ts
+++ b/src/lib/actions.ts
@@ -1,8 +1,8 @@
 import { Auth, raw, skipCSRFCheck, createActionURL } from "@auth/core"
 // @ts-expect-error Next.js does not yet correctly use the `package.json#exports` field
-import { headers as nextHeaders, cookies } from "next/headers"
+import { headers as nextHeaders, cookies } from "next/headers.js"
 // @ts-expect-error Next.js does not yet correctly use the `package.json#exports` field
-import { redirect } from "next/navigation"
+import { redirect } from "next/navigation.js"
 
 import type { NextAuthConfig } from "./index.js"
 import type { NextAuthResult, Session } from "../index.js"
diff --git a/src/lib/env.ts b/src/lib/env.ts
index 27615af273e8dbe85fa6a2de51fb6d0774b8c971..62dfdfee49879a1da452764e587f26a7fcb0468b 100644
--- a/src/lib/env.ts
+++ b/src/lib/env.ts
@@ -1,5 +1,5 @@
 // @ts-expect-error Next.js does not yet correctly use the `package.json#exports` field
-import { NextRequest } from "next/server"
+import { NextRequest } from "next/server.js"
 import type { NextAuthConfig } from "./index.js"
 import { setEnvDefaults as coreSetEnvDefaults } from "@auth/core"
 
diff --git a/src/lib/index.ts b/src/lib/index.ts
index 2223cdeaa3703e6a3cd5faee78d95af046526a75..367a54e7e9a7b0bf9cde020dbc7cbdd1565ed878 100644
--- a/src/lib/index.ts
+++ b/src/lib/index.ts
@@ -1,8 +1,8 @@
 import { Auth, createActionURL, type AuthConfig } from "@auth/core"
 // @ts-expect-error Next.js does not yet correctly use the `package.json#exports` field
-import { headers } from "next/headers"
+import { headers } from "next/headers.js"
 // @ts-expect-error Next.js does not yet correctly use the `package.json#exports` field
-import { NextResponse } from "next/server"
+import { NextResponse } from "next/server.js"
 import { reqWithEnvURL } from "./env.js"
 
 import type { AuthAction, Awaitable, Session } from "@auth/core/types"
```

**Reply by trainerkelly** [reply]
It's my pleasure! It took me a while to find a solution and because I literally could not find another solution anywhere else (I tried for several hours), that is what I came up with. I figured I wasn't the only one.

Ooo, pnpm patch!! That's awesome!! 

If anyone else has a better solution, feel free to post it lol

**Reply by lenichols** [reply]
@trainerkelly my application is doing an automated build so that nissing file /server will get recreated. any ideas of a more forma approach to a fix?

**Reply by mperper** [reply]
Thank you!

**Comment by zagzook** [comment]
I am getting the same issue here with the same error and same dependencies.  I cannot go back to NextJS 14.1.4 because then I would have to change the react version and other dependencies not an option.  the fix below the issue on that is that my errors are coming from next-auth itself there for I would have to go into the node_modules folder and add  the ,js in the files.   Not comfortable changing node_modules code.  so not an option also is there a correct fix for this or just these workarounds?

my auth,js code:
```
import NextAuth from 'next-auth';

export const { auth, handlers, signIn, signOut } = NextAuth({
  providers: [],
});
```

my dependencies:
```
  "dependencies": {
    "auth": "^1.2.3",
    "mongodb": "^6.12.0",
    "next": "^15.1.6",
    "next-auth": "^5.0.0-beta.25",
    "react": "^19.0.0",
    "react-dom": "^19.0.0"
  },
```

My Error:
```
 npx auth secret
node:internal/modules/run_main:122
    triggerUncaughtException(
    ^

Error [ERR_MODULE_NOT_FOUND]: Cannot find module 'C:\Users\ja021\Documents\nextjs_projects\codefast\node_modules\next\server' imported from C:\Users\ja021\Documents\nextjs_projects\codefast\node_modules\next-auth\lib\env.js
Did you mean to import "next/server.js"?
    at finalizeResolution (node:internal/modules/esm/resolve:275:11)
    at moduleResolve (node:internal/modules/esm/resolve:932:10)
    at defaultResolve (node:internal/modules/esm/resolve:1056:11)
    at ModuleLoader.defaultResolve (node:internal/modules/esm/loader:654:12)
    at #cachedDefaultResolve (node:internal/modules/esm/loader:603:25)
    at ModuleLoader.resolve (node:internal/modules/esm/loader:586:38)
    at ModuleLoader.getModuleJobForImport (node:internal/modules/esm/loader:242:38)
    at ModuleJob._link (node:internal/modules/esm/module_job:135:49) {
  code: 'ERR_MODULE_NOT_FOUND',
  url: 'file:///C:/Users/ja021/Documents/nextjs_projects/codefast/node_modules/next/server'
}

Node.js v22.13.0

```

**Reply by Yash5467** [reply]
downgrade to "next-auth": "^5.0.0-beta.20" works for me

**Reply by thathurtabit** [reply]
I get the same issue of `Did you mean to import "next/server.js"?` with `5.0.0-beta.20` maybe your `^` is choosing a later version? Either way only adding `.js` to the `next` imports in my `node_modules` fixes things as mentioned on this thread. Shame it's not a fix as my app will build via a docker container and will reinstall the dependencies.

**Comment by NathanNZ** [comment]
So I had this issue when starting migrating from pages to the app router. 

I moved the middleware to the app router as the documentation suggests and the error went away. As you can use both pages and app router at the same time - feel like this might be a okay fix. 

Definitely better than changing node modules!

**Reply by trainerkelly** [reply]
Oooh, that's definitely a better idea! Where does it state it in the documents? I wasn't able to find anything.

**Reply by AndreMaz** [reply]
@NathanNZ can you provide some pointers to the docs? I can't find anything

**Reply by NathanNZ** [reply]
Apologies there for the delay in reply, missed the last message. 

It's not the clearest and I'll admit I stumbled across it when debugging why the code would work in the sample but not my app and then went back to the docs to see if it was mentioned! 

'The old configuration file, contained in the API Route (pages/api/auth/[...nextauth].ts / app/api/auth/[...nextauth]/route.ts), now becomes much shorter. **These exports are designed to be used in an App Router API Route, but the rest of your app can stay in the Pages Router** if you are gradually migrating!'

https://authjs.dev/getting-started/migrating-to-v5

**Reply by AndreMaz** [reply]
@NathanNZ thank you for quick reply. Double checked on my side and I've already migrated everything to the "new" struct.

What helped me was: https://github.com/nextauthjs/next-auth/discussions/9385#discussioncomment-12023012

**Comment by KyryloSWAN** [comment]
I renamed `src/pages/api/auth/[...nextauth].ts` to `app/api/auth/[...nextauth]/route.ts` and the error went away.

"next": "15.1.6"
"next-auth": "^5.0.0-beta.25"

**Comment by oyatek** [comment]
@KyryloSWAN if I do this I get this error
```
Error: Catch-all must be the last part of the URL.
```

**Reply by oyatek** [reply]
Ah! I got it. I need to move the route out to app/!

**Comment by aeksco** [comment]
Was experiencing the same problem - updating `next.config.js` based on @KyryloSWAN suggested worked great:

```js
/**
 * Run `build` or `dev` with `SKIP_ENV_VALIDATION` to skip env validation. This is especially useful
 * for Docker builds.
 */
import "./src/env.js";

/** @type {import("next").NextConfig} */
const config = {
  reactStrictMode: true,

  // NOTE - this is a workaround for an issue with next-auth@5.0.0-beta.25 and next@15.0.0
  // @see https://github.com/nextauthjs/next-auth/discussions/10058
  transpilePackages: ["next-auth"],

  /**
   * If you are using `appDir` then you must comment the below `i18n` config out.
   *
   * @see https://github.com/vercel/next.js/issues/41980
   */
  i18n: {
    locales: ["en"],
    defaultLocale: "en",
  },
};

export default config;
```

**Reply by Awais16** [reply]
Adding next-auth to transpilePackages list resolved the issue. Thank you.
Env:
Next v15.4.4
next-auth v5.0.0-beta.29 
pnpm v10.13.1
node v22.17.1
